### DISCUSSION TITLE
REST API Parameters per Visualization

### DISCUSSION THREAD

**Discussion by sswfrd** [RevealBi/Reveal.Sdk#300] [General]
Hey guys, I have another one for you.

My current requirement is to pass REST API parameters to the individual Visualization programmatically but it must be for an individual instance of that visualization data source.

An example would be:

Visualization 1 with REST Datasource that accepts a list of Well Identifiers and a list of Trend Identifiers.

Wells: 1,5

Trends: 150, 75
 

Visualization 2 with the same REST Datasource that accepts a list of Well Identifiers and a list of Trend Identifiers.

Wells: 1,5

Trends: 300, 25

 

The end result is two visualizations that have comparison trends for two wells.  Or n number of trends and n number of wells technically.

We know you can pass values in through the header. However this does not seem like it would work as we would not want the values overwritten and we would not be able to distinguish the different instances of the REST data source.

We also know you can turn on URL parameters and have the user manually enter the values in the "Set up your data" page.

However, the user will not have the knowledge of the identifiers and plus it is cumbersome and prone to data entry errors.

Is there a way to populate these individual visualization parameters programmatically?
Thanks.

**Comment by brianlagunas** [comment]
If you set a URL on the client and set a parameter, you will be promoted for the parameter's value when creating the visualization.

Example:
![image](https://github.com/user-attachments/assets/78f69c69-e574-4ec7-8ad2-2bd504b823d0)
![image](https://github.com/user-attachments/assets/afb33ea2-625f-42bd-ab08-afa61fff9672)

The URL on the client doesn't have to be the actual URL, it can be a fake URL, as long as your server set's the data source to the proper URL in the `ChangeDataSourceAsync`.

**Reply by sswfrd** [reply]
Yes, we know you can turn on URL parameters and have the user manually enter the values in the "Set up your data" page.
However, the user will not have the knowledge of the identifiers and plus it is cumbersome and prone to data entry errors.
Is there a way to populate these individual visualization parameters programmatically?

**Reply by TorreyBetts** [reply]
@sswfrd It should be possible by creating two combinations of RVJSONDataSourceItem containing a RVRestDataSourceItem. These two would use a URL with no placeholders or parameters on the client, then they get filled in on the server within IRVDataSourceProvider. With this setup the user would see two data source item to chose from, with no setup screen to be filled in. 

I'm beginning to work on an example now, although it may take until tomorrow.

**Reply by sswfrd** [reply]
In this scenario, would we then pass these values through the header? Would they remain as parameters in the rdash file?

**Reply by TorreyBetts** [reply]
> In this scenario, would we then pass these values through the header? Would they remain as parameters in the rdash file?

The values for the URL parameters wouldn't be passed through the header. I'd expect you'd only need to pass user related information through the header in the event you want to use it with logic when filling in the parameters values on the server.

In the attached file below, I set the URL and parameters on the server. This should be enough to help you get started with your requirements. If you run into any issues, let me know.

[ExampleServerJSONParameters.zip](https://github.com/user-attachments/files/17549228/ExampleServerJSONParameters.zip)

**Reply by sswfrd** [reply]
Thank you for your help, but I don't think I clearly explained the issue.
The user can choose the same Web API several times because they are creating several Visualizations.  We do not know how many times they will choose this Web API.  In my example I chose 2 to make it easier.

The user will decide what Trends they want to visualize. We would like to create a custom UI to allow them to choose what they want and then pass those values to the Web API as parameters.  These parameters are not hard coded on the server. This custom UI is our own UI, nothing we would expect Infragistics to create.

**Reply by brianlagunas** [reply]
I'm not sure what your flow looks like or how you have your "custom dialog" implemented, but you have a couple of options. One, build them on the client like this:
```js
const restDataSourceItem = new $.ig.RVRESTDataSourceItem(restDataSource);
restDataSourceItem.parameters = { "orderId": "12345" };
const jsonDSI_1 = new $.ig.RVJsonDataSourceItem(restDataSourceItem);
jsonDSI_1.id = "JSON_1"
jsonDSI_1.title = "Option 1";

const restDataSourceItem2 = new $.ig.RVRESTDataSourceItem(restDataSource);
restDataSourceItem2.parameters = { "orderId": "67890" };
const jsonDSI_2 = new $.ig.RVJsonDataSourceItem(restDataSourceItem2);
jsonDSI_2.id = "JSON_2"
jsonDSI_2.title = "Option 2";
```

Or two, you can pass the data source id and the parameters in setAdditionalHeaders and dynamically set the params on the server.

**Reply by sswfrd** [reply]
We haven't built anything yet, looking into our options as to best present the users with the ability to pass in their own parameters.  Thank you for this suggestion. I will try out a POC.
